File: tic_tac_toe_ai.cpp
#include <iostream>
#include <vector>
using namespace std;

class TicTacToe {
private:
    static const int SIZE = 3;
    static const char EMPTY = ' ';
    static const char HUMAN = 'X';
    static const char COMPUTER = 'O';

    vector<vector<char>> board;
    bool isHumanTurn;
    bool isGameStarted;

public:
    TicTacToe() {
        board = vector<vector<char>>(SIZE, vector<char>(SIZE, ' '));
        isHumanTurn = true;
        isGameStarted = false;
        initializeBoard();
    }

    void initializeBoard() {
        char cellValue = '1';
        for (int i = 0; i < SIZE; i++) {
            for (int j = 0; j < SIZE; j++) {
                board[i][j] = cellValue++;
            }
        }
    }

    void playGame() {
        cout << "Welcome to Tic-Tac-Toe!" << endl;

        while (!isGameOver()) {
            printBoard();

            if (isHumanTurn) {
                makeHumanMove();
            } else {
                makeComputerMove();
            }

            isHumanTurn = !isHumanTurn;
            isGameStarted = true;
        }

        printBoard();
        printResult();
    }

    bool isGameOver() {
        return isBoardFull() || hasPlayerWon(HUMAN) || hasPlayerWon(COMPUTER);
    }

    bool isBoardFull() {
        for (int i = 0; i < SIZE; i++) {
            for (int j = 0; j < SIZE; j++) {
                if (isdigit(board[i][j])) {
                    return false;
                }
            }
        }
        return true;
    }

    bool hasPlayerWon(char player) {
        for (int i = 0; i < SIZE; i++) {
            if (board[i][0] == player && board[i][1] == player && board[i][2] == player) {
                return true;
            }
        }

        for (int j = 0; j < SIZE; j++) {
            if (board[0][j] == player && board[1][j] == player && board[2][j] == player) {
                return true;
            }
        }

        if (board[0][0] == player && board[1][1] == player && board[2][2] == player) {
            return true;
        }
        if (board[0][2] == player && board[1][1] == player && board[2][0] == player) {
            return true;
        }

        return false;
    }

    void printBoard() {
        cout << "-------------" << endl;
        for (int i = 0; i < SIZE; i++) {
            cout << "| ";
            for (int j = 0; j < SIZE; j++) {
                if (isHumanTurn && board[i][j] != HUMAN && board[i][j] != COMPUTER) {
                    cout << (isGameStarted ? EMPTY : board[i][j]) << " | ";
                } else {
                    cout << board[i][j] << " | ";
                }
            }
            cout << endl << "-------------" << endl;
        }
    }

    void makeHumanMove() {
        int position;
        do {
            cout << "Enter position (1-9): ";
            cin >> position;
        } while (!isValidMove(position));

        int row = (position - 1) / SIZE;
        int col = (position - 1) % SIZE;
        board[row][col] = HUMAN;
    }

    bool isValidMove(int position) {
        int row = (position - 1) / SIZE;
        int col = (position - 1) % SIZE;

        if (position < 1 || position > SIZE * SIZE) {
            cout << "Invalid move! Try again." << endl;
            return false;
        }
        if (!isdigit(board[row][col])) {
            cout << "Cell already occupied! Try again." << endl;
            return false;
        }
        return true;
    }

    void makeComputerMove() {
        vector<int> bestMove = minimax(0, COMPUTER);
        int position = bestMove[0] * SIZE + bestMove[1] + 1;
        board[bestMove[0]][bestMove[1]] = COMPUTER;
        cout << "Computer chose position: " << position << endl;
    }

    vector<int> minimax(int depth, char player) {
        vector<int> bestMove = { -1, -1, 0 };
        int bestScore;

        if (player == COMPUTER) {
            bestScore = -10;
        } else {
            bestScore = 10;
        }

        if (isGameOver()) {
            int score = evaluateBoard();
            return { -1, -1, score };
        }

        for (int i = 0; i < SIZE; i++) {
            for (int j = 0; j < SIZE; j++) {
                if (isdigit(board[i][j])) {
                    board[i][j] = player;
                    vector<int> currentMove = minimax(depth + 1, (player == COMPUTER) ? HUMAN : COMPUTER);
                    int currentScore = currentMove[2];
                    board[i][j] = board[i][j] = ('0' + i * SIZE + j + 1);

                    if (player == COMPUTER) {
                        if (currentScore > bestScore) {
                            bestScore = currentScore;
                            bestMove[0] = i;
                            bestMove[1] = j;
                        }
                    } else {
                        if (currentScore < bestScore) {
                            bestScore = currentScore;
                            bestMove[0] = i;
                            bestMove[1] = j;
                        }
                    }
                }
            }
        }

        bestMove[2] = bestScore;
        return bestMove;
    }

    int evaluateBoard() {
        if (hasPlayerWon(COMPUTER)) {
            return 1;
        } else if (hasPlayerWon(HUMAN)) {
            return -1;
        } else {
            return 0;
        }
    }

    void printResult() {
        if (hasPlayerWon(HUMAN)) {
            cout << "Congratulations! You won!" << endl;
        } else if (hasPlayerWon(COMPUTER)) {
            cout << "Computer wins!" << endl;
        } else {
            cout << "It's a draw!" << endl;
        }
    }
};

int main() {
    TicTacToe game;
    game.playGame();
    return 0;
}

File: TicTacToeAI.java
package TicTacToe;


import java.util.Scanner;


public class TicTacToeAI {
    private char [][]board;

    public char player = 'X', opponent = 'O';
    public TicTacToeAI()
    {
        board=new char [3][3];

        intializeBoard();

    }
    public void intializeBoard()
    {
        for(int i=0;i<3;i++)
        {
            for(int j=0;j<3;j++)
            {
                board[i][j]=' ';
            }
        }
    }
    public void printBoard()
    {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                System.out.print(board[i][j]);
                if (j < 2) {
                    System.out.print(" | ");
                }
            }
            System.out.println();
            if (i < 2) {
                System.out.println("---------");
            }
        }

    }
    public boolean isBoardFull()
    {

        for(int row=0;row<3;row++)
        {
            for(int col=0;col<3;col++)
            {
                if(board[row][col]==' ')
                {

                    return false;
                }
            }
        }
        return true;
    }
    public boolean isGameOver() {
        return (isGameWon(player) || isGameWon(opponent) || isBoardFull());
    }

    public boolean isGameWon(char player)
    {
        for(int row=0;row<3;row++)
        {
            if(board[row][0]==player && board[row][1]==player && board[row][2]==player)
            {
                return true;
            }
        }
        for(int col=0;col<3;col++)
        {
            if(board[0][col]==player && board[1][col]==player && board[2][col]==player)
            {
                return true;
            }
        }
        if (board[0][0] == player && board[1][1] == player && board[2][2] == player) {
            return true;
        }
        if (board[0][2] == player && board[1][1] == player && board[2][0] == player) {
            return true;
        }
        return false;
    }

    public int minimax(char currplayer,int depth)
    {
        if (isGameWon(player)) {
            return 1;
        } else if (isGameWon(opponent)) {
            return -1;
        } else if (isBoardFull()) {
            return 0;
        }

        int bestScore = (currplayer== opponent) ? Integer.MAX_VALUE : Integer.MIN_VALUE;
        for(int i=0;i<3;i++)
        {
            for(int j=0;j<3;j++)
            {
                if(board[i][j]==' ')
                {
                    board[i][j]=currplayer;
                    int score= minimax((currplayer==opponent)?player:opponent,depth+1);
                    board[i][j]=' ';
                    if (currplayer == opponent) {
                        bestScore = Math.min(score, bestScore);
                    } else {
                        bestScore = Math.max(score, bestScore);
                    }

                }
            }
        }



        return bestScore;


    }
    public Move findBestMove()
    {
        int bestScore=Integer.MAX_VALUE;
        Move bestMove=null;
        for(int i=0;i<3;i++)
        {
            for(int j=0;j<3;j++)
            {
                if(board[i][j]==' ')
                {
                    board[i][j]=opponent;
                    int score=  minimax(player,0);
                    board[i][j]=' ';
                    if(score<bestScore)
                    {
                        bestScore=score;
                        bestMove=new Move(i,j);

                    }
                }
            }

        }
        return bestMove;
    }
    public static class Move{
        int row;
        int col;
        Move(int row,int col)
        {
            this.row=row;
            this.col=col;
        }
    }

    public static void main(String[] args) {
        TicTacToeAI game=new TicTacToeAI();
        Scanner sc=new Scanner(System.in);
        while(!game.isGameOver())
        {
            game.printBoard();
            System.out.print("Enter your move (row column): ");
            int playerRow = sc.nextInt();
            int playerCol = sc.nextInt();
            if (game.board[playerRow][playerCol] == ' ') {
                game.board[playerRow][playerCol] = game.player;
            } else {
                System.out.println("Invalid move. Cell is already occupied. Try again.");
                continue;
            }
            if (game.isGameWon(game.player)) {
                game.printBoard();
                System.out.println("Congratulations! You win!");
                break;
            }
            if(game.isBoardFull())
            {
                game.printBoard();
                System.out.println("It's a draw!");
                break;
            }
            Move move=game.findBestMove();
            game.board[move.row][move.col]= game.opponent;
            System.out.println("Computer's move: " + move.row + " " + move.col);
            if (game.isGameWon(game.opponent)) {
                game.printBoard();
                System.out.println("Computer wins. Better luck next time!");
                break;
            }

        }
        if (!game.isGameWon(game.player) && !game.isGameWon(game.opponent)) {
            game.printBoard();
            System.out.println("It's a draw!");
        }
        sc.close();
    }

}
